api/src/wfl/module/covid.clj

```diff
@@ -94,82 +96,122 @@
+                     (when-not (:started workload)
+                               (patch-workload tx workload {:finished now}))
+                     (workloads/load-workload-for-id tx id))]
+             (when-not started
+                       (throw (UserException. "Cannot stop workload before it's been started."
+                                              {:workload workload})))
+             (if-not (or stopped finished) (stop! workload (utc-now)) workload)))
 
 (defn ^:private workload-to-edn [workload]
-  (-> workload
-      (util/select-non-nil-keys workload-metadata-keys)
-      (dissoc :pipeline)
-      (update :source   util/to-edn)
-      (update :executor util/to-edn)
-      (update :sink     util/to-edn)))
-
-(defoverload workloads/create-workload!    pipeline create-covid-workload)
-(defoverload workloads/start-workload!     pipeline start-covid-workload)
-(defoverload workloads/update-workload!    pipeline update-covid-workload)
-(defoverload workloads/stop-workload!      pipeline stop-covid-workload)
-(defoverload workloads/retry               pipeline batch/retry-unsupported)
-(defoverload workloads/load-workload-impl  pipeline load-covid-workload-impl)
-(defmethod   workloads/workflows           pipeline
-  [tx {:keys [executor] :as _workload}]
-  (executor/executor-workflows tx executor))
-(defmethod   workloads/workflows-by-status pipeline
-  [tx {:keys [executor] :as _workload} status]
-  (executor/executor-workflows-by-status tx executor status))
-(defoverload workloads/to-edn             pipeline workload-to-edn)
+      (-> workload
+          (util/select-non-nil-keys workload-metadata-keys)
+          (dissoc :pipeline)
+          (update :source util/to-edn)
+          (update :executor util/to-edn)
+          (update :sink util/to-edn)))
+
+(defn retry-workflows
+      [workload workflows]
+      ; Get the workflows from the terra executor details table
+      ;(->> workload :executor :details
+      ;     (format "SELECT * FROM %s")
+      ;     (jdbc/query (postgres/wfl-db-config))
+      ;     (wfl.debug/trace)
+      ;     )
+
+      ; Get the rows from the terra details table for the workflows
+      ; iterate through the results and get the distinct reference values
+      ; resubmit the reference values
+      ; write back to the table, putting the new workflow id into the retry field of the failed one
+      (let [workflow_ids (util/to-quoted-comma-separated-list (map :uuid workflows))
+            executor (get-in workload [:executor])
+            executor_details (get-in workload [:executor :details])
+            query (format "SELECT * FROM %s WHERE workflow IN (%s)" executor_details workflow_ids)
+            results (jdbc/query (postgres/wfl-db-config) [query])
+            distinct_references (set (map :reference results))]
+           (for [ref distinct_references]
+                (let [submission (executor/create-submission! executor ref)]
```

Comment:
Take a look at the source code for [`executor/create-submission!`](https://github.com/broadinstitute/wfl/blob/develop/api/src/wfl/executor.clj#L159-L164):
```
(defn ^:private create-submission!
  "Update `methodConfiguration` to use `reference`.
  Create and return submission in `workspace` for `methodConfiguration` via Firecloud."
  [{:keys [workspace methodConfiguration] :as executor} reference]
  (update-method-configuration! executor reference)
  (firecloud/submit-method workspace methodConfiguration))
```

Before submitting the method, we programmatically alter the method configuration to use the snapshot reference's *name* as its root entity type.  That takes place in [`executor/update-method-configuration!`](https://github.com/broadinstitute/wfl/blob/develop/api/src/wfl/executor.clj#L138-L157).

Note that the destructuring of the `_reference` argument gives us a hint that we expect the reference passed in to be a map like `{:name "some-snapshot-reference-name" …}`. 
```
(defn ^:private update-method-configuration!
  "Update `methodConfiguration` in `workspace` with snapshot reference `name`
  as :dataReferenceName via Firecloud.
  Update executor table record ID with incremented `methodConfigurationVersion`."
  [{:keys [id
           workspace
           methodConfiguration
           methodConfigurationVersion] :as _executor}
   {:keys [name]                       :as _reference}]
…
```

We see how the name is used, talking to Firecloud to update the method configuration:
```
…
    (firecloud/update-method-configuration
     workspace methodConfiguration
     (assoc current :dataReferenceName name :methodConfigVersion inc'd))
…
```

Where this method was originally called, we passed it a [*snapshot reference object fetched from Rawls*](https://github.com/broadinstitute/wfl/blob/develop/api/src/wfl/executor.clj#L127).  Ex.
```
{:cloningInstructions "COPY_NOTHING",
  :description "",
  :name "GP20210430",
  :reference {:instanceName "terra", :snapshot "0e308334-80f0-4cb2-9988-e57a6f67f0c5"},
  :referenceId "6374c360-b980-442c-abe4-37f144b65109",
  :referenceType "DATA_REPO_SNAPSHOT",
  :workspaceId "42f490fc-6bc6-4080-af85-dc3175e84815"}
```

What you've passed to `executor/create-submission!` and by extension to `executor/update-method-configuration!` is a *snapshot reference ID* pulled from the executor details table.  Ex.
```
"6374c360-b980-442c-abe4-37f144b65109"
```

I will put in a specific recommendation above, but my general advice is that you need to pass a snapshot reference object from Rawls and not a standalone ID.

Reply:
Something else that we may do is add a precondition to `executor/update-method-configuration!` to validate that `_reference` has a non-nil field `name`, or add some similar check.  The fact that the method configuration update happened successfully and didn't complain about the missing snapshot reference name made debugging more challenging.